ledger/complete/mtrie/trie/trie.go

```diff
@@ -101,49 +100,63 @@ func (mt *MTrie) String() string {
 	return trieStr + mt.root.FmtStr("", "")
 }
 
-// UnsafeRead read payloads for the given paths. It is called unsafe as it requires the
-// paths to be sorted
-// TODO move consistency checks from Forrest into Trie to obtain a safe, self-contained API
+// UnsafeRead read payloads for the given paths. It is called unsafe as it modifies the
+// input paths (paths are sorted when the function returns).
+// TODO move consistency checks from Forest into Trie to obtain a safe, self-contained API
 func (mt *MTrie) UnsafeRead(paths []ledger.Path) []*ledger.Payload {
-	return mt.read(mt.root, paths)
+	// allocate the result buffer
+	res := make([]*ledger.Payload, len(paths))
+	mt.read(&res, mt.root, paths)
+	return res
 }
 
-func (mt *MTrie) read(head *node.Node, paths []ledger.Path) []*ledger.Payload {
+func (mt *MTrie) read(res *[]*ledger.Payload, head *node.Node, paths []ledger.Path) {
 	// path not found
 	if head == nil {
-		res := make([]*ledger.Payload, 0, len(paths))
-		for range paths {
-			res = append(res, ledger.EmptyPayload())
+		for i := range paths {
+			(*res)[i] = ledger.EmptyPayload()
 		}
-		return res
+		return
 	}
 	// reached a leaf node
 	if head.IsLeaf() {
-		res := make([]*ledger.Payload, 0)
-		for _, p := range paths {
+		for i, p := range paths {
 			if bytes.Equal(head.Path(), p) {
-				res = append(res, head.Payload())
+				(*res)[i] = head.Payload()
 			} else {
-				res = append(res, ledger.EmptyPayload())
+				(*res)[i] = ledger.EmptyPayload()
 			}
 		}
-		return res
+		return
 	}
 
-	lpaths, rpaths := utils.SplitSortedPaths(paths, mt.height-head.Height())
+	// partition step to quick sort the paths:
+	// lpaths contains all paths that have `0` at the partitionIndex
+	// rpaths contains all paths that have `1` at the partitionIndex
+	partitionIndex := utils.SplitPaths(paths, mt.Height()-head.Height())
+	lpaths, rpaths := paths[:partitionIndex], paths[partitionIndex:]
+	lres, rres := (*res)[:partitionIndex], (*res)[partitionIndex:]
 
-	// TODO make this parallel
-	payloads := make([]*ledger.Payload, 0)
+	// read values from left and right subtrees in parallel
+	wg := sync.WaitGroup{}
+	parallelRecursionThreshold := 32 // thresold to avoid the parallelization going too deep in the recursion
 	if len(lpaths) > 0 {
```

Comment:
Can we move this check to the top?
```
if len(paths) {
  return nil
}
```

And then we can remove `if len(lpaths) > 0` and `if len(rpaths) > 0`

Reply:
Yep, good idea! I didn't update the original code because this change gives better readability but costs extra recursive calls. I think readability is better here 👌